packages/preferences/src/store/actions.js

```diff
@@ -47,3 +47,36 @@ export function setDefaults( scope, defaults ) {
 		defaults,
 	};
 }
+
+/** @typedef {() => Promise<Object>} WPPreferencesPersistenceLayerGet */
+/** @typedef {(*) => void} WPPreferencesPersistenceLayerSet */
+/**
+ * @typedef WPPreferencesPersistenceLayer
+ *
+ * @property {WPPreferencesPersistenceLayerGet} get An async function that gets data from the persistence layer.
+ * @property {WPPreferencesPersistenceLayerSet} set A  function that sets data in the persistence layer.
+ */
+
+/**
+ * Sets the persistence layer.
+ *
+ * When a persistence layer is set, the preferences store will:
+ * - call `get` immediately and update the store state to the value returned.
+ * - call `set` with all preferences whenever a preference changes value.
+ *
+ * `setPersistenceLayer` should ideally be dispatched at the start of an
+ * application's lifecycle, before any other actions have been dispatched to
+ * the preferences store.
+ *
+ * @param {WPPreferencesPersistenceLayer} persistenceLayer The persistence layer.
+ *
+ * @return {Object} Action object.
+ */
+export async function setPersistenceLayer( persistenceLayer ) {
+	const persistedData = await persistenceLayer.get();
```

Comment:
I did think about that. I don't think it's worth it, at least for this first iteration.

For the UI to be responsive, `preferences.get` needs to be synchronous. 

The store could be updated with server data in the background, but then there would need to be some handling of data consistency. If a user performs an action that updates a preference while a call to `persistenceLayer.get()` is in flight, the user's optimistic action should overwrite the in flight API request.

Ideally I think it would be a push system (server updates push to the store just like user actions) instead of a pull system, and then this would be easier. Because of that it might be something to try when exploring collaborative editing, as there will likely be some need for that kind of server infrastructure.